### PR TITLE
modeling assignmentOperations with api-version 2018-11-01

### PR DESCRIPTION
**Pull request by haitch**
### Changes
 - Modeling assignmentOperations, so SDK client can retrieve details about any assignment and history as well.

### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.

### CODE REVIEW

**Comment by haitch** [comment]
@annatisch can you take a look? 
We didn't maintain the go sdk, probably some code gen robot need to adjust.

**Comment by annatisch** [comment]
Thanks @haitch!
We will need to get the Go code gen working successfully in order to merge. The error looks transient so I will try a re-run - otherwise I will get some Go experts to help us out :)

Adding @ravbhatnagar for ARM review.

Could you please take a look at the model validation error?
```
 error : 
operationId: AssignmentOperations_List
scenario: Assignment
source: response
responseCode: '200'
errorDetails:
  code: INVALID_TYPE
  params:
    - object
    - array
  message: Expected type object but found type array
  title: /definitions/AssignmentOperationList
  description: List of AssignmentOperation.
  url: >-
    specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
```

**Comment by haitch** [comment]
@annatisch the issue has been fixed, please take a look again. 
Thanks

**Comment by annatisch** [comment]
Thanks @haitch - there's still a model validation error.
Looks like a typo - `"nextLink"` needs a capital 'L'
```
 error : 
operationId: AssignmentOperations_List
scenario: Assignment
source: response
responseCode: '200'
errorDetails:
  code: OBJECT_ADDITIONAL_PROPERTIES
  params:
    - - nextlink
  message: 'Additional properties not allowed: nextlink'
  title: /definitions/AssignmentOperationList
  description: List of AssignmentOperation.
  url: >-
    specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
```

**Comment by annatisch** [comment]
Hi @haitch - in addition, I see that you have configuration for Python SDK generation, however no reference to the repo - could you please add an extra line here in order to have the Python SDK generate:
https://github.com/Azure/azure-rest-api-specs/tree/b0a026b7604e3996531d21fd3581bb3a12285730/specification/blueprint/resource-manager#swagger-to-sdk

You just need to add:
```
  - repo: azure-sdk-for-python
```

**Comment by jhendrixMSFT** [comment]
@AutorestCI regenerate azure-sdk-for-go

**Comment by annatisch** [comment]
Ping @ravbhatnagar for ARM review :)

File: specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
Comment (RyanBensonMSFT) [inline specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json]: Are any of these properties readonly? 
If they are computed in the RP, mark them as readonly so the CLI doesn't require them as input.

File: specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
Comment (RyanBensonMSFT) [inline specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json]: Are any of these (kind or jobState for example) Enums? If so you should declare the possible values here:

Example:
        "intervalLength": {
          "type": "string",
          "description": "Interval value in minutes used to create LogAnalytics call rate logs.",
          "enum": [
            "ThreeMins",
            "FiveMins",
            "ThirtyMins",
            "SixtyMins"
          ],
          "x-ms-enum": {
            "name": "IntervalInMins",
            "modelAsString": false
          }
        }

File: specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
Comment (RyanBensonMSFT) [inline specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json]: Can this be modeled as a child action or resourceType? 
The URI would look something like:
/{scope}/providers/Microsoft.Blueprint/blueprintAssignments/{assignmentName}/operations/{assignmentOperationName}/deployments

File: specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
Comment (RyanBensonMSFT) [inline specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json]: This also seems like a good candidate to be modeled as a child action or resource too.
The proposed model assumes the user wants all deployments and the full history of each deployment when retrieving an operation. This gets a lot more cumbersome when they call the list all operations API.

Resource group deployments have a similar scenario and have split these concepts into different nested resources. The deployment contains high level aggregated info and a child resource 'operations' exists for details.

**Review by ryansbenson — CHANGES_REQUESTED** [review]

File: specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
Comment (haitch) [reply specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json]: this is Get/List only, I don't have the readonly problem.

File: specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
Comment (haitch) [reply specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json]: no, 
1. assignmentOperation is already a child resource, with result aggregation.
2. in our backend, we store AssignmentOperation as a whole unit.
3. the size of deploymentJob is pretty small, just some metadata, no huge payload.
4. it's easier to consume.

File: specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
Comment (haitch) [reply specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json]: in most of the case, it should just have Result, history only appears when you have retry.

each individual retry just have error and resourceCreated, also the resourceCreated would just have {id,Name,Type}, rest of the property will be trimed off.

File: specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json
Comment (haitch) [reply specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/assignmentOperation.json]: model as enum will force us to upgrade api-version on enum changes.

Will re-consider model them as enum when we are more stable and ready to GA.

**Comment by ravbhatnagar** [comment]
As per discussion with Marcin over email, here is what will happen. tracking it here- https://github.com/Azure/azure-rest-api-specs/issues/4903
/assignmentOperations
In a resource ID, it will look like this /subscriptions/{id}/providers/Microsoft.Blueprint/blueprintAssignments/{name}/assignmentOperations, which is redundant. On the upside the generated SDK already calls this object AssignmentOperations, so it will match the API now. The downside is that we will continue supporting /operations on API version 2017-11-11-*. API versions 2018-11-01-* and later would use /assignmentOperations. This will remain in our RP manifest and service code until 2017-11-11-preview is fully deprecated and removed. This seems like a good compromise to get this PR unblocked.

**Comment by ravbhatnagar** [comment]
Signing off from ARM side.

**Comment by haitch** [comment]
#4903 is addressed in [latest commit](https://github.com/Azure/azure-rest-api-specs/pull/4590/commits/1208b90b526f3872c07ae9d9a91646102b94e0a3)

**Comment by haitch** [comment]
@annatisch  can you take a look again and merge?

**Comment by annatisch** [comment]
Thanks @haitch - model validation pass is still failing:
```
 error :
operationId: AssignmentOperations_List
errorDetails:
  code: DOUBLE_FORWARD_SLASHES_IN_URL
  message: >-
    In operation "AssignmentOperations_List", example for parameter "scope":
    "/subscriptions/f8df94f2-2f5a-4f4a-bcaf-1bb992fb564b" starts with a forward
    slash and the path template:
"/{scope}/providers/Microsoft.Blueprint/blueprintAssignments/{assignmentName}/assignmentOperations"    contains a forward slash before the parameter starts. This will cause double
    forward slashes  in the request url. Thus making it incorrect. Please
    rectify the example.

 error : 
operationId: AssignmentOperations_Get
errorDetails:
  code: DOUBLE_FORWARD_SLASHES_IN_URL
  message: >-
    In operation "AssignmentOperations_Get", example for parameter "scope":
    "/subscriptions/f8df94f2-2f5a-4f4a-bcaf-1bb992fb564b" starts with a forward
    slash and the path template:
"/{scope}/providers/Microsoft.Blueprint/blueprintAssignments/{assignmentName}/assignmentOperations/{assignmentOperationName}"
    contains a forward slash before the parameter starts. This will cause double
    forward slashes  in the request url. Thus making it incorrect. Please
    rectify the example.
```

**Comment by haitch** [comment]
fixed, @annatisch can you approve?

**Comment by annatisch** [comment]
Thanks @haitch - I will try to review this afternoon

**Comment by annatisch** [comment]
@haitch - is this API already deployed/available?

**Comment by haitch** [comment]
it was deployed with /operations last week.

I've also updated our RP with the rename, the rename deploy soon.

**Comment by annatisch** [comment]
Thanks - please update this thread when the rename has been deployed (or when deployment has started).

**Review by annatisch — COMMENTED** [review]
Thanks @haitch - the PR looks good! :)
My only comments are minor regarding the consistent capitalization of the descriptions.

**Review by annatisch — APPROVED** [review]
LGTM.
Waiting on service deployment to merge :)

**Comment by annatisch** [comment]
@haitch - Is this API now public?

**Comment by haitch** [comment]
@annatisch it is public, you can safely merge this PR.

**Comment by annatisch** [comment]
@RyanBensonMSFT - could you please sign off your review if your requested changes have been addressed?

Adding @lmazuel to merge PR once that sign off happens.

**Comment by filizt** [comment]
@RyanBensonMSFT please let us know if anything is blocking the sign off.

**Review by ryansbenson — APPROVED** [review]

**Comment by filizt** [comment]
@lmazuel could you merge the PR please?

**Pull request merged**
